crates/sui-types/src/transaction.rs

```diff
@@ -2054,6 +2036,22 @@ impl TransactionData {
     }
 }
 
+pub fn is_system_tx(kind: &TransactionKind, sender: SuiAddress) -> bool {
+    // Keep this as an exhaustive match so that we can't forget to update it.
+    match kind {
+        TransactionKind::ChangeEpoch(_)
+        | TransactionKind::Genesis(_)
+        | TransactionKind::ConsensusCommitPrologue(_)
+        | TransactionKind::ConsensusCommitPrologueV2(_)
+        | TransactionKind::ConsensusCommitPrologueV3(_)
+        | TransactionKind::ConsensusCommitPrologueV4(_)
+        | TransactionKind::AuthenticatorStateUpdate(_)
+        | TransactionKind::RandomnessStateUpdate(_)
+        | TransactionKind::EndOfEpochTransaction(_) => true,
+        TransactionKind::ProgrammableTransaction(_) => sender == SuiAddress::default(),
```

Comment:
I think this would be tricky and likely not ideal.
For settlement transactions, I think we should use a native system transaction in data format, which contains all necessary information, and at execution time, we construct a PTB to do the actual settlement.
The benefits of that is it still allow us to to cleanly see that this is a special system transaction, as well as makes it easy to bypass some of the mutability restrictions (e.g. when we need to settle in parallel in multiple transactions)

Reply:
As discussed in person, I think the mutability restrictions need a more systematic approach anyway.

We could go with the new TransactionKind approach, but consider:
- `sender = 0x0` is a perfectly clean way to detect system txns, its just a change from what we've done before
- adding a new TransactionKind has a bigger blast radius because client and indexing code has to change - these changes can get quite large and also disruptive for indexers